### PR TITLE
eir: add UEFI boot protocol

### PR DESCRIPTION
**Pull request by marv7000**
This PR adds UEFI as a boot protocol to eir. It's architecture independent and introduces a "protos" directory for future independent protocols.

### CODE REVIEW

**Comment by no92** [comment]
Depends on managarm/frigg#62.

File: kernel/eir/meson.build [hunk 1]
```diff
@@ -1,8 +1,11 @@
 eir_generic_sources = files(
 	'../common/libc.cpp',
 	'../common/font-8x16.cpp',
-	'generic/main.cpp',
-	'generic/debug.cpp')
+	'generic/main.cpp')
+
+if not build_uefi
+	eir_generic_sources += files('generic/debug.cpp')
+endif
```
Comment (avdgrinten) [inline kernel/eir/meson.build:8]: Instead of overriding `debug.cpp`, `debug.cpp` should offer a way for the protocol to install a custom logging sink. For example, by simply storing a global `void (*logUsingBootProto)(const char *)`.

File: kernel/eir/meson.build [hunk 2]
```diff
@@ -13,6 +16,22 @@ eir_includes = [include_directories(
...
+	eir_dependencies = [frigg, libarch]
+	if arch == 'aarch64'
+		subdir('arch/arm')
+	elif arch == 'riscv64'
+		subdir('arch/riscv')
+	elif arch == 'x86_64'
+		subdir('arch/x86')
+	endif
+	subdir('protos/uefi')
+	subdir_done()
```
Comment (avdgrinten) [inline kernel/eir/meson.build:32]: Why is this done instead of continuing with the remainder of the meson.build?

File: kernel/eir/protos/uefi/eir-internal/arch/types.hpp [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#pragma once
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/eir-internal/arch/types.hpp:1]: An `arch/` header should not be in `protos/`.

File: kernel/memory.txt [hunk 4]
```diff
@@ -31,11 +31,6 @@ FFFF'F000'0000'0000
 	Log ring buffer memory
 	Referenced in eir/main.cpp, thor/generic/core.cpp
 
-FFFF'F000'0000'0000
-	Length: 256MiB
-	Log ring buffer memory
-	Referenced in eir/main.cpp, thor/generic/core.cpp
-
```
Comment (avdgrinten) [inline kernel/memory.txt:38]: I assume that this is outdated?

File: meson_options.txt [hunk 5]
```diff
@@ -1,15 +1,20 @@
...
+    type : 'boolean',
     value : false
 )
 
-option('build_drivers', 
-    type : 'boolean', 
+option('build_uefi',
+    type : 'boolean',
     value : false
 )
```
Comment (avdgrinten) [inline meson_options.txt:9]: Can't we deduce this from the target triple?

File: meson.build [hunk 6]
```diff
@@ -44,7 +45,7 @@ if build_kernel
 	add_project_link_arguments(['-Wl,--orphan-handling=warn', '-Wl,-znoexecstack'], language: ['c', 'cpp'])
 endif
 
-if build_kernel or build_drivers
+if build_kernel or build_uefi or build_drivers
 	add_languages('c', 'cpp', native: false)
 endif
```
Comment (avdgrinten) [inline meson.build:50]: A lot of the meson.build logic can be simplified if we assume that `build_kernel` is set whenever `build_uefi` is set.

File: kernel/eir/protos/uefi/meson.build [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	'freestnd-c-hdrs-' + host_machine.cpu_family(),
+	required: true,
+	fallback: ['freestnd-c-hdrs', 'freestnd_c_hdrs_dep'],
+)
+
+freestnd_cxx_hdrs_dep = dependency(
+	'freestnd-cxx-hdrs-' + host_machine.cpu_family(),
+	required: true,
+	fallback: ['freestnd-cxx-hdrs', 'freestnd_cxx_hdrs_dep'],
+)
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/meson.build:17]: Shouldn't eir in general use these dependencies (and not just the uefi target)?

**Review by avdgrinten — COMMENTED** [review]
Reviewed the build changes. I did not review the EFI code yet.

File: kernel/eir/arch/arm/load64.S [hunk 8]
```diff
@@ -30,4 +30,6 @@ eirEnterKernel:
 	mov x0, x4
 	br x2
 
+#ifndef __clang__
 	.section .note.GNU-stack,"",%progbits
+#endif
```
Comment (avdgrinten) [inline kernel/eir/arch/arm/load64.S:35]: Why is this gated behind non-Clang instead of non-UEFI?

File: kernel/eir/arch/x86/arch.cpp [hunk 9]
```diff
@@ -178,6 +178,12 @@ void initProcessorEarly() {
 	// Keep in sync with the SMP trampoline in thor.
 	uint64_t pat = 0x00'00'01'00'00'00'04'06;
 	common::x86::wrmsr(0x277, pat);
+
+	// Clear Write Protect bit.
+	uint64_t cr0;
+	asm volatile ("mov %%cr0, %0" : "=r" (cr0));
+	cr0 &= ~(1 << 16);
+	asm volatile ("mov %0, %%cr0" : : "r" (cr0));
```
Comment (avdgrinten) [inline kernel/eir/arch/x86/arch.cpp:186]: Why is this necessary?

File: kernel/eir/arch/x86/load64.S [hunk 10]
```diff
@@ -22,8 +22,11 @@ eirEnterKernel:
 	or $0x00010000, %rax # Enable WP
 	mov %rax, %cr0
 
+	cli
```
Comment (avdgrinten) [inline kernel/eir/arch/x86/load64.S:25]: Do UEFI apps run with interrupts enabled? What if an interrupt happens before this `cli`?

File: kernel/eir/protos/uefi/entry.cpp [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+#include "helpers.hpp"
+
+static_assert(sizeof(char16_t) == sizeof(wchar_t), "Strings are not UTF-16-ish, are you missing -fshort-wchar?");
+extern "C" void eirEnterKernel(uintptr_t, uint64_t, uint64_t) __attribute__((sysv_abi));
+
+namespace eir {
+
+const efi_system_table *st = nullptr;
+const efi_boot_services *bs = nullptr;
+efi_handle handle = nullptr;
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/entry.cpp:20]: Are these variables local to the file? If yes, they should be in an anonymous namespace. Otherwise, it would be good to pick more descriptive names.

File: kernel/eir/protos/uefi/entry.cpp [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+	frg::parse_arguments(cmdLine, args);
+
+	eir::infoLogger() << "eir: command line='" << cmdLine << "'" << frg::endlog;
+
+	// this needs to be volatile as GDB sets this to true on attach
+	volatile bool __eir_gdb_ready = eir_gdb_ready_val;
+
+	if(!eir_gdb_ready_val) {
+		constexpr arch::scalar_register<uint8_t> offset{0};
+		auto port = arch::global_io.subspace(0xCB7);
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/entry.cpp:78]: Please add a comment what this magic port number does.

File: kernel/eir/protos/uefi/entry.cpp [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+	efi_file_protocol *kernelFile = nullptr;
+	EFI_CHECK(fsOpen(&kernelFile, asciiToUcs2(thor_path)));
+	auto kernel_info = fsGetInfo(kernelFile);
+	efi_physical_addr thor = 0;
+	EFI_CHECK(bs->allocate_pages(AllocateAnyPages, EfiLoaderData, (kernel_info->file_size >> 12) + 1, &thor));
+	fsRead(kernelFile, kernel_info->file_size, 0, thor);
+
+	efi_file_protocol *initrdFile = nullptr;
+	EFI_CHECK(fsOpen(&initrdFile, asciiToUcs2(initrd_path)));
+	efi_file_info *initrdInfo = fsGetInfo(initrdFile);
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/entry.cpp:108]: In the future, we want to store `thor` inside the initrd CPIO. Does it make sense to implement this model right away? CC @qookei

File: kernel/eir/protos/uefi/entry.cpp [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+	EFI_CHECK(bs->allocate_pages(AllocateAnyPages, EfiLoaderData, (initrdInfo->file_size >> 12) + 1, &initrd));
+	EFI_CHECK(fsRead(initrdFile, initrdInfo->file_size, 0, initrd));
+
+	// Get the frame buffer.
+	efi_graphics_output_protocol *gop = nullptr;
+
+	{
+		efi_guid gop_protocol = EFI_GRAPHICS_OUTPUT_PROTOCOL_GUID;
+		EFI_CHECK(bs->locate_protocol(&gop_protocol, nullptr, reinterpret_cast<void **>(&gop)));
+		assert(gop->mode->info->version == 0);
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/entry.cpp:121]: We should not use `assert()` to check for runtime errors.

File: kernel/eir/protos/uefi/entry.cpp [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+	// Exit boot services.
+	EFI_CHECK(bs->exit_boot_services(handle, mapKey));
+	bs = nullptr;
+
+	frg::array<InitialRegion, 4> reservedRegions = {};
+	size_t nReservedRegions = 0;
+
+	reservedRegions[nReservedRegions++] = {reinterpret_cast<uintptr_t>(loadedImage->image_base), loadedImage->image_size};
+	reservedRegions[nReservedRegions++] = {reinterpret_cast<uintptr_t>(thor), kernel_info->file_size};
+	reservedRegions[nReservedRegions++] = {reinterpret_cast<uintptr_t>(initrd), initrdInfo->file_size};
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/entry.cpp:162]: What about the stuff that is allocated above? In particular, `cmdLine`, the memory map and allocations in helpers.cpp?

File: kernel/eir/protos/uefi/helpers.hpp [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+void EFI_CHECK(efi_status s, std::source_location loc = std::source_location::current());
+
+namespace eir {
+
+extern const efi_system_table *st;
+extern const efi_boot_services *bs;
+extern efi_handle handle;
+
+extern bool __eir_log_debugcon;
+extern bool __eir_log_e9;
```
Comment (avdgrinten) [inline kernel/eir/protos/uefi/helpers.hpp:20]: Why do we have double underscore here?

File: kernel/eir/arch/arm/load64.S (hunk 8 above)
Comment (marv7000) [reply kernel/eir/arch/arm/load64.S:35]: That syntax isn't supported in Clang's integrated assembler, and LLVM defaults to `-z noexecstack`. It is unrelated to UEFI.

File: kernel/eir/meson.build (hunk 2 above)
Comment (marv7000) [reply kernel/eir/meson.build:32]: IIRC libgcc or cxxshim caused build failures and/or are incompatible with the UEFI target.

Comment (avdgrinten) [reply kernel/eir/meson.build:32]: Then these libraries should be guarded by `if not build_uefi`. The `subdir_done()` also skips `KERNEL_KASAN` and `KERNEL_LOG_ALLOCATIONS` and disables a bunch of other meson vars that architectures may want to use.

File: kernel/eir/arch/x86/arch.cpp (hunk 9 above)
Comment (no92) [reply kernel/eir/arch/x86/arch.cpp:186]: We did this to work around something before, but this isn't necessary any more AFAICT.

File: kernel/eir/arch/x86/load64.S (hunk 10 above)
Comment (no92) [reply kernel/eir/arch/x86/load64.S:25]: We do this because thor assumes it's getting launched with the interrupt flag cleared. UEFI runs with it enabled, though. From the spec, revision 2.10A, page 28:

> Interrupts are enabled-though no interrupt services are supported other than the UEFI boot services timer functions (All loaded device drivers are serviced synchronously by “polling.”)

File: kernel/eir/protos/uefi/entry.cpp (hunk 11 above)
Comment (no92) [reply kernel/eir/protos/uefi/entry.cpp:20]: They are global, and calling them `st` or `gST` is very common.

File: kernel/eir/protos/uefi/entry.cpp (hunk 12 above)
Comment (no92) [reply kernel/eir/protos/uefi/entry.cpp:78]: Note to self: this port is used for exfiltrating the base address of `eir-uefi` so that it can be communicated to `vm-util.py`'s `gdb --uefi` subcommand.

File: kernel/eir/protos/uefi/entry.cpp (hunk 14 above)
Comment (no92) [reply kernel/eir/protos/uefi/entry.cpp:121]: Thing about this is that we have no way of handling an updated modeinfo struct correctly, if that were to ever happen. The spec currently only defines version 0, so asserting seems like a fair choice?

Comment (avdgrinten) [reply kernel/eir/protos/uefi/entry.cpp:121]: I don't have a problem with hard failures, but we should use `panicLogger()` and not `assert()` in these cases. `assert()` should be used to check for contract violations, not runtime errors.

Comment (avdgrinten) [reply kernel/eir/protos/uefi/entry.cpp:121]: Note that this also applies to other `assert()`s in the same file (not all of them though, for example `assert()`ing that `cmdLine` is non-null is fine as the code above is supposed to guarantee this condition.

File: meson_options.txt (hunk 5 above)
Comment (no92) [reply meson_options.txt:9]: We could check whether the target triplet's OS part is `windows` or `uefi`, but that seems a bit more complicated and prone to break than just having a meson_option for it.

Comment (marv7000) [reply meson_options.txt:9]: What about a new option for which protocol to build for? Currently, all archs assume they are always using e.g. mb2. With an extra option you could pass e.g. `uefi`, `mb2` or `limine`.

File: kernel/eir/protos/uefi/entry.cpp (hunk 13 above)
Comment (no92) [reply kernel/eir/protos/uefi/entry.cpp:128]: As discussed, this will be implemented later down the line.

**Comment by no92** [comment]
Rebased the PR, addressed the comments raised.

**Review by avdgrinten — APPROVED** [review]

**Review by qookei — APPROVED** [review]
